RTSStateMachine.py

```diff
@@ -0,0 +1,588 @@
+                    self.resume_to_previous()
+                    print(f"\nResumed to previous state")
+                    print(f"Current state: {self.current_state}")
+                    break
+                elif user_input == "3":
+                    self.advance_to_next_in_cycle()
+                    break
+                elif user_input == "4":
+                    print("Exiting system...")
+                    sys.exit()
+            except (EOFError, KeyboardInterrupt):
+                print("\n\nReceived interrupt signal. System remains paused.")
+                print("Type 1, 2, 3, or 4 to continue.")
+
+    def advance(self):
+        """Advance to the next chip position on the tray."""
+        if self.current_chip_index < len(self.chip_positions['col']) - 1:
+            self.current_chip_index += 1
+        else:
+            raise StopIteration("Reached the end of the tray.")
+
+    def get_position(self):
+        """Get the current chip position on the tray."""
+        if not self.chip_positions['col'] or self.current_chip_index >= len(self.chip_positions['col']):
+            return (0, 0)
+        return (self.chip_positions['col'][self.current_chip_index], 
+                self.chip_positions['row'][self.current_chip_index])
+
+    def advance_chip_position(self):
+        """Advance to the next chip position with automatic reset at end."""
+        try:
+            self.advance()
+            print(f"Advanced to chip position: {self.get_position()}")
+        except StopIteration:
+            print("Reached the end of the tray.")
+            self.current_chip_index = 0
+
+    def reset_tray_position(self):
+        """Reset the chip position to the beginning of the tray."""
+        self.current_chip_index = 0
+        print("Reset chip position to (1, 1)")
+
+    def is_tray_complete(self):
+        """Check if we've processed all positions on the tray."""
+        return self.current_chip_index == len(self.chip_positions['col']) - 1
+
+    def run_full_cycle(self):
+        """Run a complete test cycle for one chip and advance position."""
+        print(f"Starting full cycle at position {self.get_position()}")
+        for i in range(6):
+            self.cycle()
+        print("Full cycle complete, advancing chip position")
+        self.advance_chip_position()
+    
+    def handle_tray(self):
+        """Process all chips on the tray with full test cycles."""
+        num_chips = len(self.chip_positions['col'])
+        for i in range(num_chips):
+            print(f"\n--- Processing chip {i+1}/{num_chips} ---")
+            self.run_full_cycle()
+        print(f"\nTray processing complete! Processed {num_chips} chips.")
+
+    def get_current_chip_data(self):
+        """Get all data for the current chip as a dictionary."""
+        return {key: self.chip_positions[key][self.current_chip_index] for key in self.chip_positions}
+    
+    def set_chip_data(self, index, col=None, row=None):
```

Comment:
Can you include a definition for index (possibly all inputs) into this function?